daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/LayerTransformer.scala

```diff
@@ -282,4 +234,234 @@ case class LayerNotEnoughDataException(layerSRD: SequenceRuntimeData, state: PSt
   override def isError = true
   override def modeName = "Parse"
-  }
\ No newline at end of file
+  }
+
+
+
+class LayerRuntimeInfoImpl(val srd: SequenceRuntimeData,
+  maybeLayerCharsetEv: Maybe[LayerCharsetEv],
+  maybeLayerLengthKind: Maybe[LayerLengthKind],
+  maybeLayerLengthEv: Maybe[LayerLengthEv],
+  maybeLayerLengthUnits: Maybe[LayerLengthUnits],
+  maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv])
+  extends LayerRuntimeInfo {
+
+  def evaluatables: Seq[Evaluatable[AnyRef]] =
+    maybeLayerCharsetEv.toScalaOption.toSeq ++
+      maybeLayerLengthEv.toScalaOption.toSeq ++
+      maybeLayerBoundaryMarkEv.toScalaOption.toSeq
+
+
+  def layerInfo(state: ParseOrUnparseState): LayerInfo =
+    new LayerInfoImpl(state, srd,
+      maybeLayerCharsetEv,
+      maybeLayerLengthKind,
+      maybeLayerLengthEv,
+      maybeLayerLengthUnits,
+      maybeLayerBoundaryMarkEv)
+}
+
+/**
+ * Allows access to all the layer properties, if defined, including
+ * evaluating expressions if the properties values are defined as expressions.
+ */
+sealed trait LayerInfo {
+  def optLayerCharset: Option[java.nio.charset.Charset]
+  def optLayerLengthKind: Option[LayerLengthKind]
+  def optLayerLength: Option[Long]
+  def optLayerLengthUnits: Option[LayerLengthUnits]
+  def optLayerBoundaryMark: Option[String]
+}
+
+class LayerInfoImpl(state: ParseOrUnparseState,
+  srd: SequenceRuntimeData,
+  maybeLayerCharsetEv: Maybe[LayerCharsetEv],
+  maybeLayerLengthKind: Maybe[LayerLengthKind],
+  maybeLayerLengthEv: Maybe[LayerLengthEv],
+  maybeLayerLengthUnits: Maybe[LayerLengthUnits],
+  maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv])
+extends LayerInfo {
+
+  override def optLayerCharset: Option[Charset] = maybeLayerCharsetEv.toScalaOption.map {
+    _.evaluate(state)
+  } match {
+    case Some(bitsCharsetJava: BitsCharsetJava) => Some(bitsCharsetJava.javaCharset)
+    case None => None
+    case Some(other) => None // only java charsets are supported.
+  }
+
+  override def optLayerLengthKind: Option[LayerLengthKind] = maybeLayerLengthKind.toScalaOption
+
+  override def optLayerLength: Option[Long] = maybeLayerLengthEv.toScalaOption.map{ _.evaluate(state) }
+
+  override def optLayerLengthUnits: Option[LayerLengthUnits] = maybeLayerLengthUnits.toScalaOption
+
+  override def optLayerBoundaryMark: Option[String] = maybeLayerBoundaryMarkEv.toScalaOption.map{ _.evaluate(state) }
+}
+
+
+abstract class ByteBufferExplicitLengthLayerTransform[T](
+  layerRuntimeInfoArg: LayerRuntimeInfo,
+  layerName: String,
+  variablesNamespace: String,
+  variablesPreferredNamespacePrefix: String,
+  localNamesAndTypesOfVariablesToRead: Seq[(String, String)],
+  localNameOfVariableToWrite: String,
+  outputVariableTypeLocalNameAsString: String)
+  extends LayerTransformer(layerName, layerRuntimeInfoArg) {
+
+  /**
+   * Override to specify the length exactly.
+   */
+  protected def explicitLength() = -1L
```

Comment:
Should we remove this default implementation? Seems like -1 wouldn't never make sense and a user must implement this function.

Reply:
This is only needed for a layer transform with a built-in constant length, so that the schema author need not specify dfdlx:layerLengthKind 'explicit' and dfdl:layerLength='N' for the right length N. 

If a layer transform does use those things, then they don't need to override this method, and the -1 is ignored. 

But I agree we should just make this pure virtual. 

I think I will also rename it to layerBuiltInConstantLengthValue()